arch/x86/core/fatal.c

```diff
@@ -432,3 +432,18 @@ void z_x86_do_kernel_oops(const z_arch_esf_t *esf)
 
 	z_x86_fatal_error(reason, esf);
 }
+
+/**
+ * NMI handling
+ *
+ * Handlers should be registered with NMI_HANDLER_DEFINE(name, func, dev)
+ */
+
+void z_x86_do_kernel_nmi(const z_arch_esf_t *esf)
+{
+	Z_STRUCT_SECTION_FOREACH(nmi_handler, handler) {
```

Comment:
Without more information on why we need a special section with chained NMI handlers, I'd rather this was simpler, just a weak function which by default calls z_x86_fatal_error(), can be overridden at the application level.

In particular the need for multiple NMI handlers seems really unlikely to me, but if you have supporting use-cases do share. Keep in mind that these functions basically have to be their own standalone environment with no kernel services available. That includes driver APIs which use locks for synchronization. It looks suspicious to me that these handlers take a struct device as an argument...you have no idea what state the device is in

Reply:
@andrewboie , I have done as you requested, to make "iterable section of callback pointers at build time", is this correct?

Reply:
this doesn't answer my question: what are the use-cases for having multiple callbacks? what are you planning to do inside these callbacks? why do we need multiple callbacks?

also in this current design, if no NMI handlers are installed the exception isn't treated as a fatal error

Reply:
I need only one callback, I can make it working only for me with weak function, if somebody need another callback he would reimplement this. I will fix no callback case.

Reply:
@andrewboie,  Re-implemented with weak function.